cvxpy/lin_ops/canon_backend.py

```diff
@@ -826,6 +827,290 @@ def get_param_tensor(self, shape: tuple[int, ...], parameter_id: int) \
+
+            def parametrized_mul(x):
+                assert len(x) == 1
+                return {k: v @ x[0] for k, v in stacked_lhs.items()}
+
+            func = parametrized_mul
+        else:
+            assert len(lhs) == 1
+            reps = view.rows // lhs[0].shape[-1]
+            stacked_lhs = (np.kron(np.eye(reps), lhs[0]))
+
+            def func(x):
+                return stacked_lhs @ x
+
+        return view.accumulate_over_variables(func, is_param_free_function=is_param_free_lhs)
+
+    @staticmethod
+    def promote(lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
+        num_entries = int(np.prod(lin.shape))
+
+        def func(x):
+            # Fast way of repeating sparse matrix along axis 0
+            # See comment in https://stackoverflow.com/a/50759652
+            return x[:, np.zeros(num_entries, dtype=int), :]
+
+        view.apply_all(func)
+        return view
+
+    def mul_elem(self, lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
+        lhs, is_param_free_lhs = self.get_constant_data(lin.data, view, column=True)
+        if isinstance(lhs, dict):
+            def parametrized_mul(x):
+                assert len(x) == 1
+                return {k: v * x[0] for k, v in lhs.items()}
+
+            func = parametrized_mul
+        else:
+            def func(x):
+                return lhs[0] * x
+        return view.accumulate_over_variables(func, is_param_free_function=is_param_free_lhs)
+
+    @staticmethod
+    def sum_entries(_lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
+        def func(x):
+            return x.sum(axis=(0, 1), keepdims=True)
+
+        view.apply_all(func)
+        return view
+
+    def div(self, lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
+        lhs, is_param_free_lhs = self.get_constant_data(lin.data, view, column=True)
+        assert is_param_free_lhs
+        assert lhs.shape[0] == 1
+        data = lhs.flatten(order='F')
+
+        # dtype is important here, will do integer division if data is of dtype "int" otherwise.
+        data = np.reciprocal(data, where=data != 0, dtype=float)
+        lhs = data.reshape(lhs.shape)
+
+        def div_func(x):
+            return lhs * x
+
+        return view.accumulate_over_variables(div_func, is_param_free_function=is_param_free_lhs)
+
+    @staticmethod
+    def diag_vec(lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
+        assert lin.shape[0] == lin.shape[1]
+        k = lin.data
+        rows = lin.shape[0]
+        total_rows = int(lin.shape[0] ** 2)
+
+        def func(x):
+            shape = list(x.shape)
+            shape[1] = total_rows
+            if k == 0:
+                new_rows = np.arange(rows) * rows + np.arange(rows)
+            elif k > 0:
+                new_rows = np.arange(rows) * rows + np.arange(rows) + rows * k
+            else:
+                new_rows = np.arange(rows) * rows + np.arange(rows) - k
+            new_rows = new_rows[:rows - k]
```

Comment:
I am not sure if this would work with a negative offset, because we might be trying to slice an index that is greater than the length of new_rows. 
I am also not sure how diag_vec with a negative offset looks like? Could you provide an example, or perhaps add another unit test for that?

Reply:
As mentioned, we already check that `k` is not out of bounds when creating the `diag_mat` atom:
https://github.com/cvxpy/cvxpy/blob/245defa66c1403cc2c5929cdc8a1a5265b9a73ea/cvxpy/atoms/affine/diag.py#L51

Reply:
Ok, thanks for the pointer